Self-review

busykai opened the pull request.

busykai committed:
Add stub shim-systemd support.

Does nothing at this point.

busykai committed:
Add partial implementation.

Bootloaders are installed, but not kernels. Kernels are installed
(copied) into wrong location by a completely different piece of code.

busykai committed:
Ask bootloader where to copy the kernels.

busykai commented on configure.ac:
```diff
@@ -42,21 +42,28 @@ AC_ARG_WITH([bootloader], AS_HELP_STRING([--with-bootloader=BOOTLOADER],
         AC_MSG_CHECKING([whether an EFI bootloader is available])
         if test -e "/usr/lib/systemd/boot/efi"; then
-                bootloader="systemd-boot"
+                if test "/usr/lib/shim"; then
```
test -e

busykai commented on src/bootman/kernel.c:
```diff
@@ -592,23 +592,18 @@ bool boot_manager_install_kernel_internal(const BootManager *manager, const Kern
-        bool is_uefi = false;
-        autofree(char) *efi_boot_dir = NULL;
+        int is_uefi = (manager->bootloader->get_capabilities(manager) & BOOTLOADER_CAP_UEFI);
+        autofree(char) *efi_boot_dir = manager->bootloader->get_kernel_dst(manager);
```
check if `is_uefi`

busykai committed:
Clean up.

busykai committed:
Use get_kernel_dst() instead of hard-coded value.

busykai committed:
Add working impl of kernel install.

busykai committed:
WIP: efi boot var implementation.

busykai committed:
Get rid of the maybe uninitialized warning.

busykai committed:
Add draft implementation of boot var support.

busykai committed:
Rename files, clean up.

busykai committed:
Add/correct copyright notice.

bryteise commented on configure.ac:
```diff
@@ -42,21 +42,28 @@ AC_ARG_WITH([bootloader], AS_HELP_STRING([--with-bootloader=BOOTLOADER],
         else
                 AC_MSG_WARN([Cannot find a valid EFI bootloader, defaulting to systemd-boot])
-                bootloader="systemd-boot"
+                bootloader="shim-systemd-boot"
```
I'd guess keeping the default would be fine here.

busykai replied:
We will no longer use standalone `systemd-boot`, so it's not OK if we accidentally build a previous version. Ikey was going to have a single implementation (replacing `systemd-boot` with `shim-systemd-boot`) when I asked him what his approach would be.

bryteise replied:
Okay for me then.

bryteise commented on src/bootloaders/shim-systemd.c:
```diff
@@ -101,6 +103,10 @@ static char *shim_dst;
 static char *systemd_dst;
 
+static char *shim_systemd_get_kernel_dst(const BootManager *manager) {
+        return strdup(KERNEL_DST_DIR);
```
Is there a reason why systemd and shim implementation can't both use the destination? (basically just thinking this logic could be the same function between sd_class types and shim).

bryteise commented on src/bootloaders/shim-systemd.c:
```diff
@@ -110,7 +111,7 @@ static char *shim_systemd_get_kernel_dst(const BootManager *manager) {
 static bool shim_systemd_install_kernel(const BootManager *manager, const Kernel *kernel) {
         fprintf(stderr, "Call %s\n", __func__);
-        return true;
+        return sd_class_install_kernel_impl(manager, kernel, shim_systemd_get_kernel_dst, NULL);
```
Similar for this commit, it would be nice if the kernel install parts could be the same for sd_class and shim. I say this mostly as it seems easier to manage one rather than two locations and it doesn't really make sense to me to have a single distro with two kernels, one for secure boot and one for without.

busykai replied:
I do not have a strong opinion, but I do have an opinion.

1. I've defined layout in such a way that everything that we touch is under root `/org.clearlinux`.  We don't go any other place and there's no confusion as to where are the files that we manage.  systemd's hard-coded `/loader` is unfortunate exception, but it's life (and systemd, too).

2. Conceptually, each bootloader's implementation in `src/bootloader` is the owner of its own layout. It's not only tied to kernels, but also bootloaders themselves, their names and location. I do not think that a part of a booting scheme should be defined elsewhere. I've reused creation of conf entry for systemd-boot.

These were my reasons for doing it this way. Now, we can make the naming schemes coincide. However, I do not see how it is helpful.

Just to make sure we're on the same page: this is a replacement scheme for the current one. We will not be using them in parallel. Therefore, we will be no two location to manage. It's not that we're going to be maintaining two layouts depending on whether secure boot is enabled or not. But...

There's been a change in layout (in June, I guess) and kernels were moved. That change simply left stray kernels in their previous location. Which is not nice.

If your concern behind the location of the kernels is managing the transition, then we can make sure there's a clean up code in place that will ensure no artifacts of the previous layout are left on the ESP (i.e. kernels and bootloaders are removed).

So, I'm suggesting to let a bootloader (or bootscheme) auto-contain the definition of the layout. Let it all be in `/org.clearlinux`. Additional item as the result of this discussion would be adding code that cleans up the ESP. Would that be OK by you?

bryteise replied:
So I guess I'm thinking about folks who would like to use Clear Linux without secure boot but perhaps that doesn't make any difference in how they will still be using shim and the rest of the new bits (I wasn't really thinking about that). If that's the case as long as our migration plan is going to not leave bits behind it is okay with me with your approach.

busykai replied:
Yes, people who will not have Secure Boot enabled will be using the exact same EFI binaries (even though they will be signed) and the exact same boot path.

I'll make sure the new bootloader looks up and cleans up after the previous booting schemes.

bryteise commented on src/efi/efisupport.c:
```diff
@@ -110,11 +119,75 @@ static int find_free_boot_rec(void) {
+
+/* Given the location of the booloader, returns the partition information needed
+ * to create boot variable which points to that bootloader. */
+int get_part_info(const char *path, part_info_t *pi) {
```
I wonder if Ikey might have some ideas on reusing existing code that does this kind of thing already. I'm not really familiar with this part of cbm though.

busykai replied:
I took a look at it. [`probe.h`](https://github.com/busykai/clr-boot-manager/blob/master/src/lib/probe.h) is a high-level API quite specific to kernel installation. there's also [`blkid_stub.h`](https://github.com/busykai/clr-boot-manager/blob/master/src/lib/blkid_stub.h) which is just an indirection to `libblkid` (for testing, I assume). I might end up reusing the latter for the testing sake. In some cases (e.g. `devnode_to_devpath`), vanilla `libblkid` impl of the same is more comprehensive.

P.S. I think you commented on renamed files and the comments appear as outdated. :)

busykai committed:
Clean up.

busykai committed:
Separate host and ESP paths and use them properly.

busykai committed:
Make internal function static.

busykai committed:
Avoid creating duplicate Boot* vars.

Attempt to find variables with exaact same content before creating a new
one.

busykai committed:
Clean up. Refactor.

busykai committed:
Move implementation. Rename and prefix functions.

busykai committed:
Remove debug output.

busykai committed:
Reuse set_default_kernel from systemd-class.

busykai committed:
Make bootvar_create return boot var name.

Also, pass the ESP mount point instead of host bootloader path to
determine the partition information.

busykai committed:
Use capital hex in boot vars(compliant with UEFI).

busykai committed:
Fix memory usage.

busykai committed:
Add the boot record first to the boot order.
